App/Services/RpcController.cs

```diff
@@ -251,6 +256,9 @@ private void MutateState(Action<RpcModel> mutator)
         using (_stateLock.Lock())
         {
             mutator(_state);
+            // Unset the startup progress if the VpnLifecycle is not Starting
+            if (_state.VpnLifecycle != VpnLifecycle.Starting)
+                _state.VpnStartupProgress = null;
```

Comment:
It smells funny to me that we need this. Why is the VPN service sending these updates if the VPN is not starting?

Reply:
The progress updates are sent separately to the `StartResponse` message. They can't be included in the `StartResponse` as we don't allow sending multiple replies or streaming data in the protocol (although you can in gRPC)

This handles clearing the `VpnStartupProgress` when the VPN transitions from `starting` to `started`. It also handles the case if we receive progress messages out of order (or when the client doesn't think the VPN is `starting`).

Reply:
>  It also handles the case if we receive progress messages out of order (which shouldn't be possible unless there's a bug), or when the client doesn't think the VPN is starting.

This is the bit that smelled funny to me. We generally don't want to respond to bugs (or possible bugs) in one part of the application by building workarounds in another. We want to fix the bugs and/or find ways to reduce the risk of them occurring.

> This handles clearing the VpnStartupProgress when the VPN transitions from starting to started.

This is legit. I think it would be better, though, to make it part of the `RpcModel` itself, since it's an _invariant_ of the model and we'd expect it to apply for all users, not just the RpcController.

Reply:
Sorted this out with custom getters and setters in the model